---
title: "CCBR1155 CS033351 CUT&RUN Analysis"
author: "Samantha Sevilla"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Background
######################################################################
## Project info
######################################################################
This analysis report includes:

| CS ID | Type | Species | CCBR | Notes | 
| -- | -- | -- | -- | -- |
| CS033351 | Cut and Run | Hg38 | 1155 |  |

######################################################################
## Links to Tracks
######################################################################
USBC Genome Tracks were created for each project. They can be accessed here:

- [CS033351](https://hpcnihapps.cit.nih.gov/genome/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr14%3A37589552%2D37595249&hgsid=138465_n0aQkE0ISyBdKa6Mueyvwz2kLbwa)

# Analysis
######################################################################
## Notes
######################################################################
- Deduplicated peaks were identified, using the following parameters:

  - fragments based method
  
  - normalized, relaxed method
  
- Peaks were then annotated with gene ID's. 

  - Comparisons including ALL genes include the corresponding gene of any signficant peaks identified (p value <0.05, fc +/- 1.5)
  
  - Peak annotations were stratified into those that were over represented, as well as those under-represented between each comparison. For example peakID chr1:100099186-100105441 had the following average counts for these two samples:	CRISPR_NC_2_Smyd3 (~65,287) and CRISPR_KO_10_Smyd3 (~349,940). The log2FoldChange between these two samples is approximately -2.29. The comparison CRISPR_NC_Smyd3_vs_KO_Smyd3 with a -2.29 log2FC would indicate that there is an UNDER-representation in CRISPR_NC_2_Smyd3 as compared to CRISPR_KO_10_Smyd3 for this gene ID (Cntnap5b).

```{r paths_lists, echo=FALSE, warning=FALSE, include=FALSE}
############################################################
# project information
############################################################
#paths
parent_dir="~/../../Volumes/CUTRUN/"
input_dir = "~/../../Volumes/CUTRUN/analysis/CS033351/"
output_car_dir = "~/../../Volumes/CUTRUN/analysis/CS033351/r_analysis_230104/"
output_dir=output_car_dir
car_subpath=paste0(input_dir,"carlisle_221231/results/peaks/contrasts/")
output_pic_dir=paste0(output_car_dir,"img")
dir.create(file.path(output_car_dir), showWarnings = FALSE)
dir.create(file.path(output_pic_dir), showWarnings = FALSE)
cs_id="CS033351"
```

```{r user_params}
############################################################
# set global variables 
############################################################
padj_cutoff = 0.05
species = "Homo Sapiens"
log2fc_cutoff_car=0.3785 #FC of 1.3
method="fragments" #AUC or fragments
dedup_status="dedup"
scalesfbymean="Y"
rawcountsprescaled="N"

# reference info
if (species=="Homo Sapiens"){
  txdb_pkg="TxDb.Hsapiens.UCSC.hg38.knownGene"
}

# group to ignore
ignore_groups=c()

```

```{r pkg, echo=FALSE, warning=FALSE, include=FALSE}
# package list
list.of.packages=c("DESeq2","edgeR", "tidyverse","dplyr", "DT","reshape2","pander","plotly", "ggplot2", "ggfortify", "ggrepel", "yaml", "EnhancedVolcano", "RUVSeq", "ChIPseeker", "stringr","RColorBrewer","HTSFilter","pheatmap",txdb_pkg, "UpSetR","ggupset","ggimage")

#install as needed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) BiocManager::install(new.packages)

# source functions
source(paste0(parent_dir,'github/CAR_core_functions.R'))

# load packages
invisible(lapply(list.of.packages, library, character.only = TRUE))

#negate
`%ni%` <- Negate(`%in%`)

# assign annodbs
if (species=="Homo Sapiens"){
  txdb<-TxDb.Hsapiens.UCSC.hg38.knownGene
  annodb="org.Hs.eg.db"
}
```

## **Data Processing**
```{r contrast, include=FALSE,warning=FALSE, message=FALSE}
####################
# run metadata input
####################
#groups input
groups_df = read.csv(paste0(input_dir,"groups.tab"),sep="\t",header = FALSE)[,c("V1","V2")]
colnames(groups_df) = c("sampleid","group")
rownames(groups_df)=groups_df$sampleid

## if there is a subset gene list, select files
gene_file = paste0(parent_dir,"docs/merged_gene_list_v2.csv")
pi_gene_df = read.csv(gene_file)

# read in the contrast list
contrast_df=read.csv(paste0(input_dir,"contrasts.tsv"),sep="\t")
colnames(contrast_df)=c("CNTRL","TREATMENT")
contrast_df$contrast=paste0(contrast_df$CNTRL,"_vs_",contrast_df$TREATMENT)

#filter
groups_df=subset(groups_df, group %ni% ignore_groups)
contrast_df=subset(contrast_df, (CNTRL %in% unique(groups_df$group)) | (TREATMENT %in% unique(groups_df$group)))
```

The following samples are included in this analysis:
```{r dt_samp, echo=FALSE,warning=FALSE, message=FALSE}
DT::datatable(groups_df)
```

The following contrasts are included in this analysis:
```{r dt_group, echo=FALSE,warning=FALSE, message=FALSE}
DT::datatable(contrast_df)
```

## **Quality Control**
Raw counts are read in and filtered by counts per million reads (CPM). Filtering thresholds are set such that at least two samples must have a CPM of >0.5 to be reviewed. Two plots were generated for each sample: Relative log expression (RLE) plots and Principal Coordinate Analysis (PCA) plots.

Reviewing sample QC allows us to determine if there are outliers present in analysis and the success of data normalization. If there are, samples are removed in the "Selected Samples Normalized" plots.

## Quality Control  {.tabset}

#### HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min
```{r qc1, echo=FALSE, warning=FALSE, message=FALSE}
exclusion_list=c()
contrast_id="HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min"
norm_type_cutandrun="narrowGo_peaks"

peak_anno_1=main_prep_qc_core(contrast_id,exclusion_list)
```

#### HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min
```{r qc2, echo=FALSE, warning=FALSE, message=FALSE}
exclusion_list=c()
contrast_id="HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min"
norm_type_cutandrun="broadGo_peaks"

peak_anno_1=main_prep_qc_core(contrast_id,exclusion_list)
```

#### HA_Mock_IFNb_antiHA_2min_vs_HA_UHRF1_IFNb_antiHA_2min
```{r qc3, echo=FALSE, warning=FALSE, message=FALSE}
exclusion_list=c()
contrast_id="HA_Mock_IFNb_antiHA_2min_vs_HA_UHRF1_IFNb_antiHA_2min"
norm_type_cutandrun="narrowGo_peaks"

peak_anno_2=main_prep_qc_core(contrast_id,exclusion_list)
```

#### HA_Mock_IFNb_antiHA_2min_vs_HA_UHRF1_IFNb_antiHA_2min
```{r qc4, echo=FALSE, warning=FALSE, message=FALSE}
exclusion_list=c()
contrast_id="HA_Mock_IFNb_antiHA_2min_vs_HA_UHRF1_IFNb_antiHA_2min"
norm_type_cutandrun="broadGo_peaks"

peak_anno_2=main_prep_qc_core(contrast_id,exclusion_list)
```

## **Data Processing**
Reviewing counts of significant peaks guides analysis parameters downstream.

```{r collapsedf, echo=FALSE, warning=FALSE, message=FALSE}
#read gene list
gene_list=read.csv(paste0(parent_dir,"/docs/merged_gene_list_v2.csv"),sep=",")

##########################################################################################
# create significant gene lists
contrast_id="HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min"
norm_type_cutandrun="broadGo_peaks"
merged_sig_df=create_sig_df(contrast_id,norm_type_cutandrun)

contrast_id="HA_Mock_IFNb_antiHA_2min_vs_HA_UHRF1_IFNb_antiHA_2min"
norm_type_cutandrun="narrowGo_peaks"
merged_sig_df=rbind(merged_sig_df,
                        create_sig_df(contrast_id,norm_type_cutandrun))

# collapse counts
if (nrow(merged_sig_df)>0){
  collapsed_df=create_sig_collapsed_df(merged_sig_df)
}
##########################################################################################
# create nonsignificant gene lists
contrast_id="HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min"
norm_type_cutandrun="broadGo_peaks"
merged_nonsig_df=create_nonsig_df(contrast_id,norm_type_cutandrun)

contrast_id="HA_Mock_IFNb_antiHA_2min_vs_HA_UHRF1_IFNb_antiHA_2min"
norm_type_cutandrun="narrowGo_peaks"
merged_nonsig_df=rbind(merged_nonsig_df,
                        create_nonsig_df(contrast_id,norm_type_cutandrun))

# collapse counts
collapsed_df=create_nonsig_collapsed_df(merged_nonsig_df)
```

Review summary annotation for all peaks.

### HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min

```{r graphics1,echo=FALSE, warning=FALSE, message=FALSE}
chipseeker_plots(peak_anno_1)
```

### HA_UHRF1_IFNb_antiHA_2min_vs_HA_Mock_IFNb_antiHA_2min

```{r graphics2,echo=FALSE, warning=FALSE, message=FALSE}
chipseeker_plots(peak_anno_2)
```

## **Volcano Plots**

Volcano plots are created to show the statistical significance (FDR) versus magnitude of change (fold change).

### HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min {.tabset}
```{r, v1, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
contrast_id="HA_Mock_IFNb_antiHA_1min_vs_HA_UHRF1_IFNb_antiHA_1min"
generate_volcano_plots(data_type="CAR",
                       contrast_id,
                       gene_list_in="OFF")
```

### HA_UHRF1_IFNb_antiHA_2min_vs_HA_Mock_IFNb_antiHA_2min {.tabset}
```{r, v2, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
contrast_id="HA_UHRF1_IFNb_antiHA_2min_vs_HA_Mock_IFNb_antiHA_2min"
generate_volcano_plots(data_type="CAR",
                       contrast_id,
                       gene_list_in="OFF")
```