---
title: "CCBR1155 CS031014 CUT&RUN Analysis"
author: "Samantha Sevilla"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---
## Project info
This analysis report comparison two projects

| CS ID | Type | Species | CCBR | Notes | 
| -- | -- | -- | -- | -- |
| CS031014 | Cut and Run | Hg38 | 1155 |  |

## Links
- [CS031014](https://hpcnihapps.cit.nih.gov/genome/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chrX%3A15560138%2D15602945&hgsid=138042_bQaCEQ730iKs2v633AtElgyIVxFO)

```{r paths_lists, echo=FALSE, warning=FALSE, include=FALSE}
############################################################
# project information
############################################################
#paths
parent_dir="~/../../Volumes/CUTRUN/"
input_dir = "~/../../Volumes/CUTRUN/analysis/CS031014/"
output_car_dir = "~/../../Volumes/CUTRUN/analysis/CS031014/r_analysis_230118/"
output_dir=output_car_dir
car_subpath=paste0(input_dir,"carlisle_221206/results/peaks/contrasts/")
img_dir=paste0(output_car_dir,"img/")
dir.create(file.path(output_car_dir), showWarnings = FALSE)
dir.create(file.path(img_dir), showWarnings = FALSE)
cs_id="CS031014"
```

```{r user_params}
############################################################
# set global variables 
############################################################
padj_cutoff = 0.05
species = "Homo Sapiens"
log2fc_cutoff_car=0.3785 #FC of 1.3
method="fragments" #AUC or fragments
dedup_status="dedup"
scalesfbymean="Y"
rawcountsprescaled="N"

# reference info
if (species=="Homo Sapiens"){
  txdb_pkg="TxDb.Hsapiens.UCSC.hg38.knownGene"
}

# group to ignore
ignore_groups=c()

```

```{r pkg, echo=FALSE, warning=FALSE, include=FALSE}
# package list
list.of.packages=c("DESeq2","edgeR", "tidyverse","dplyr", "DT","reshape2","pander","plotly", "ggplot2", "ggfortify", "ggrepel", "yaml", "EnhancedVolcano", "RUVSeq", "ChIPseeker", "stringr","RColorBrewer","HTSFilter","pheatmap",txdb_pkg, "UpSetR","ggupset","ggimage","ggpubr","plyr")

#install as needed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) BiocManager::install(new.packages)

# source functions
source(paste0(parent_dir,'github/CAR_core_functions.R'))

# load packages
invisible(lapply(list.of.packages, library, character.only = TRUE))

#negate
`%ni%` <- Negate(`%in%`)

# assign annodbs
if (species=="Homo Sapiens"){
  txdb<-TxDb.Hsapiens.UCSC.hg38.knownGene
  annodb="org.Hs.eg.db"
}
```

## **Data Processing**
```{r contrast, include=FALSE,warning=FALSE, message=FALSE}
####################
# run metadata input
####################
#groups input
groups_df = read.csv(paste0(input_dir,"groups.tab"),sep="\t",header = FALSE)[,c("V1","V2")]
colnames(groups_df) = c("sampleid","group")
rownames(groups_df)=groups_df$sampleid

## if there is a subset gene list, select files
gene_file = paste0(parent_dir,"docs/merged_gene_list_v2.csv")
pi_gene_df = read.csv(gene_file)

# read in the contrast list
contrast_df=read.csv(paste0(input_dir,"contrasts.tsv"),sep="\t")
colnames(contrast_df)=c("CNTRL","TREATMENT")
contrast_df$contrast=paste0(contrast_df$CNTRL,"_vs_",contrast_df$TREATMENT)

#filter
groups_df=subset(groups_df, group %ni% ignore_groups)
contrast_df=subset(contrast_df, (CNTRL %in% unique(groups_df$group)) | (TREATMENT %in% unique(groups_df$group)))
```

The following samples are included in this analysis:
```{r dt_samp, echo=FALSE,warning=FALSE, message=FALSE}
DT::datatable(groups_df)
```

The following contrasts are included in this analysis:
```{r dt_group, echo=FALSE,warning=FALSE, message=FALSE}
DT::datatable(contrast_df)
```

## **Quality Control**
Raw counts are read in and filtered by counts per million reads (CPM). Filtering thresholds are set such that at least two samples must have a CPM of >0.5 to be reviewed. Two plots were generated for each sample: Relative log expression (RLE) plots and Principal Coordinate Analysis (PCA) plots.

Reviewing sample QC allows us to determine if there are outliers present in analysis and the success of data normalization. If there are, samples are removed in the "Selected Samples Normalized" plots.

## Quality Control  {.tabset}

#### 53_H3K4me3_vs_HN6_H3K4me3
```{r qc1, echo=FALSE, warning=FALSE, message=FALSE}
exclusion_list=c()
contrast_id="53_H3K4me3_vs_HN6_H3K4me3"
norm_type_cutandrun="narrowGo_peaks"

peak_anno_4=main_prep_qc_core(contrast_id,exclusion_list)
```

#### 53_H4K20m3_vs_HN6_H4K20me3
```{r qc2, echo=FALSE, warning=FALSE, message=FALSE}
exclusion_list=c()
contrast_id="53_H4K20m3_vs_HN6_H4K20me3"
norm_type_cutandrun="broadGo_peaks"

peak_anno_20=main_prep_qc_core(contrast_id,exclusion_list)
```

## **Data Processing**
Reviewing counts of significant peaks guides analysis parameters downstream.

```{r collapsedf, echo=FALSE, warning=FALSE, message=FALSE}
#read gene list
gene_list=read.csv(paste0(parent_dir,"/docs/merged_gene_list_v2.csv"),sep=",")

##########################################################################################
# create significant gene lists
contrast_id="53_H3K4me3_vs_HN6_H3K4me3"
norm_type_cutandrun="narrowGo_peaks"
merged_sig_df=create_sig_df(contrast_id,norm_type_cutandrun)

contrast_id="53_H4K20m3_vs_HN6_H4K20me3"
norm_type_cutandrun="broadGo_peaks"
merged_sig_df=rbind(merged_sig_df,
                        create_sig_df(contrast_id,norm_type_cutandrun))

# collapse counts
collapsed_df=create_sig_collapsed_df(merged_sig_df)

##########################################################################################
# create nonsignificant gene lists
contrast_id="53_H3K4me3_vs_HN6_H3K4me3"
norm_type_cutandrun="narrowGo_peaks"
merged_nonsig_df=create_nonsig_df(contrast_id,norm_type_cutandrun)

contrast_id="53_H4K20m3_vs_HN6_H4K20me3"
norm_type_cutandrun="broadGo_peaks"
merged_nonsig_df=rbind(merged_nonsig_df,
                        create_nonsig_df(contrast_id,norm_type_cutandrun))

# collapse counts
collapsed_df=create_nonsig_collapsed_df(merged_nonsig_df)
```

Review summary annotation for all peaks.

### 53_H3K4me3_vs_HN6_H3K4me3

```{r graphics1,echo=FALSE, warning=FALSE, message=FALSE}
chipseeker_plots(peak_anno_4)
```

### 53_H4K20m3_vs_HN6_H4K20me3

```{r graphics2,echo=FALSE, warning=FALSE, message=FALSE}
chipseeker_plots(peak_anno_20)
```

## **Volcano Plots**

Volcano plots are created to show the statistical significance (FDR) versus magnitude of change (fold change).

### 53_H3K4me3_vs_HN6_H3K4me3 {.tabset}
```{r, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
contrast_id="53_H3K4me3_vs_HN6_H3K4me3"
generate_volcano_plots(data_type="CAR",
                       contrast_id,
                       gene_list_in="OFF")
```

### 53_H4K20m3_vs_HN6_H4K20me3 {.tabset}
```{r, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
contrast_id="53_H4K20m3_vs_HN6_H4K20me3"
generate_volcano_plots(data_type="CAR",
                       contrast_id,
                       gene_list_in="OFF")
```


## *Intensity Boxplots*
```{r}
contrast_id="53_H3K4me3_vs_HN6_H3K4me3"
sample_list=c("HN6_H3K4me3_1","HN6_H3K4me3_2","HN6_H3K4me3_3",
              "53_H3K4me3_1","53_H3K4me3_2","53_H3K4me3_3")
color_pal=c("black","black","black","blue","blue","blue")
generate_intensity_boxplot(contrast_id,
                           sample_list,
                           color_pal=color_pal)

contrast_id="53_H3K4me3_vs_HN6_H3K4me3"
sample_list=c("HN6_H3K4me3_2",
              "53_H3K4me3_3")
color_pal=c("black","blue")
generate_intensity_boxplot(contrast_id,
                           sample_list,
                           color_pal=color_pal)

```

