---
title: "CCBR1155 CS028891 RNA Analysis"
author: "Samantha Sevilla"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---

## Project Info
| CS ID | Type | Species | CCBR | Notes | 
| -- | -- | -- | -- | -- |
| CS028891 | RNASeq | Hg38 | 1155 | |

## Links
- [CS028891](https://hpcnihapps.cit.nih.gov/genome/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chr1%3A228424845%2D228425360&hgsid=138045_YaxblMavQd49PFnKV3dfLnT9Icne)

```{r setup, include=FALSE}
#dirs
parent_dir="~/../../../Volumes/CUTRUN/"
input_dir = "~/../../../Volumes/CUTRUN/analysis/CS028891/pipeliner_220713/"
output_rna_dir = "~/../../../Volumes/CUTRUN/analysis/CS028891/r_analysis_230323_newannotation/"
output_dir=output_rna_dir

img_dir=paste0(output_rna_dir,"img/")
dir.create(file.path(output_rna_dir), showWarnings = FALSE)
dir.create(file.path(img_dir), showWarnings = FALSE)
```

```{r thresholds}
# group to ignore
#ignore_groups=c("CRISPR_52_with_IFNb","CRISPR_53_with_IFNb","parental_HN6_with_IFNb")
ignore_groups=""

#cntrl
cntrl="parental_HN6_with_IFNb"
csid="CS028891"

# Threshold values
log2fc_cutoff=0.3785 #FC of 1.3
padj_cutoff=0.1
species_in = "Homo sapiens" #either "Homo sapiens" or "Mus musculus"
analysis_type = "DESeq2" #either limma or DESeq2
sample_consensus_threshold=0.1
read_minimum_threshold=10

gene_bodies_list=c("Promoter","Intron","Exon","5'UTR","3'UTR")
gene_bodies_filename=paste(unlist(substr(gene_bodies_list, start = 1, stop = 1)),collapse="")

#databases to include in analysis options include:
#http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp
#"C1","C2:BIOCARTA","C2:KEGG","C2:REACTOME","C2:WIKIPATHWAYS","C5:MF","C5:BP","C5:CC","H"
db_list=c("C2:KEGG","C2:REACTOME","C2:WIKIPATHWAYS","C5:MF","C5:BP","C5:CC")

```

```{r pkg, echo=FALSE, warning=FALSE, include=FALSE}
#species db
if (species_in == "Homo sapiens"){
  species_db="org.Hs.eg.db"
} else if (species_in == "Mus musculus"){
  species_db="org.Mm.eg.db"
} else{
    print("Species not approved")
}

# package list
list.of.packages=c("BiocManager","tidyr","DT","RUVSeq","RColorBrewer","textshape",
                   "pheatmap","grid","dplyr","EnhancedVolcano","edgeR","DESeq2","ggplot2","plotly",
                   "msigdbr","clusterProfiler","ggpubr","ggridges","factoextra","gplots","ReactomePA",species_db)

#install as needed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) BiocManager::install(new.packages)

# source functions
source(paste0(parent_dir,'github/CAR_core_functions.R'))

# load packages
invisible(lapply(list.of.packages, library, character.only = TRUE))

#negate
`%ni%` <- Negate(`%in%`)
```


## **Data Processing**

```{r contrast, include=FALSE}
####################
# run metadata input
####################
#groups input
groups_df = read.csv(paste0(input_dir,"groups.tab"),sep="\t",header = FALSE)[,c("V1","V2")]
colnames(groups_df) = c("sampleid","group")
rownames(groups_df)=groups_df$sampleid

## if there is a subset gene list, select files
gene_file = paste0(parent_dir,"docs/gene_list_master.csv")
gene_df = read.csv(gene_file)

# read in the contrast list
contrast_df=read.csv(paste0(input_dir,"contrasts.tab"),header=FALSE,sep="\t")
contrast_df$V1=gsub("  "," ",contrast_df$V1)
contrast_df$V1=gsub("  "," ",contrast_df$V1)
colnames(contrast_df)=c("CNTRL","TREATMENT")
contrast_df$contrast=paste0(contrast_df$CNTRL,"_vs_",contrast_df$TREATMENT)

#filter
groups_df=subset(groups_df, group %ni% ignore_groups)
contrast_df=subset(contrast_df, (CNTRL %in% unique(groups_df$group)) | (TREATMENT %in% unique(groups_df$group)))
```

The following samples are included in this analysis:
```{r dt_samp, echo=FALSE}
DT::datatable(groups_df)
```

The following contrasts are including in this analysis:
```{r dt_group, echo=FALSE}
DT::datatable(contrast_df)
```

Raw counts are read in and filtered by counts per million reads (CPM). Filtering thresholds are set such that at least two samples must have a CPM of >0.5 to be reviewed. Two plots were generated for each sample: Relative log expression (RLE) plots and Principal Coordinate Analysis (PCA) plots.

- Upper quantile normalization was attempted. This provided some correction, however, variance is still not near 0 in some samples.

- Finally, DESEQ2 normalization was attempted. This provided the best correction and was selected as the normalization technique.

```{r qc, echo=FALSE, warning=FALSE, message=FALSE}
main_prep_qc_rna_core(flag_type="pre_norm")

main_prep_qc_rna_core(flag_type="upper")

main_prep_qc_rna_core(flag_type="DESEQ2")
```

```{r}
# create sig df
sigpath=paste0("~/../../Volumes/CUTRUN/analysis/CS029758/r_analysis_230321_newannotations/",
                 "contrast_sig_53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb_PIE53.csv")
create_sig_RNA_df(contrast_id="CRISPR_53_with_IFNb-parental_HN6_with_IFNb",
                       car_sig_path=sigpath)
create_sig_RNA_df(contrast_id="CRISPR_52_with_IFNb-parental_HN6_with_IFNb",
                       car_sig_path=sigpath)
```

## *Sample heatmaps* 
```{r}
contrast_id="CRISPR_53_with_IFNb-parental_HN6_with_IFNb"

# ALL GENES
sigpath=paste0("~/../../Volumes/CUTRUN/analysis/CS029758/r_analysis_230223/",
             "contrast_sig_53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb_PIE53.csv")
sample_sub_list=c("parental_HN6_with_IFNb_2","parental_HN6_with_IFNb_3",
                  "CRISPR_53_with_IFNb_1","CRISPR_53_with_IFNb_3")
generate_replicate_heatmaps_rna(contrast_id=contrast_id,
                            scale_flag="ON",
                            sample_subset=sample_sub_list,
                            car_sig_path=sigpath,
                            gene_list_name="")

# subsets of gene lists
gene_list_options=c("APM_INFA")
for (genelistID in gene_list_options){
  generate_replicate_heatmaps_rna(contrast_id=contrast_id,
                                  scale_flag="ON",
                                  sample_subset=sample_sub_list,
                                  car_sig_path=sigpath,
                                  gene_list_name=genelistID)
}
```


## **Volcano Plots**

Volcano plots are created to show the statistical significance (FDR) versus magnitude of change (fold change).

### CRISPR_52_with_IFNb {.tabset}
```{r, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_volcano_plots(data_type="RNA",
                       contrast_id="CRISPR_52_with_IFNb-parental_HN6_with_IFNb",
                       gene_list_name=c("APM","IFNalpha"))
```

### CRISPR_53_with_IFNb {.tabset}
```{r, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_volcano_plots(data_type="RNA",
                       contrast_id="CRISPR_53_with_IFNb-parental_HN6_with_IFNb",
                       gene_list_name=c("APM","IFNalpha"))
```

## **Heatmaps**

```{r heatmap, echo=FALSE,warning=FALSE, message=FALSE}
####################
# run heatmaps
###################
# find significant genes
treatment="CRISPR_52_with_IFNb"
merged_fc=create_sig_gene_df(cntrl,treatment,25)

treatment="CRISPR_53_with_IFNb"
merged_fc=full_join(merged_fc,create_sig_gene_df(cntrl,treatment,25))

# fill in NA's in gene list
merged_fc=fillin_sig_gene_df(merged_fc)

#subset for logfc only
heatmap_df=create_heatmap_df(merged_fc,"_with_IFNb")

# create output df
create_output_df(merged_fc,25,"_with_IFNb")
```


## **Enriched Pathways** 

Pathway analysis was performed using several publicly available databases (http://www.gsea-msigdb.org/gsea/msigdb/collections.jsp). For each database, dotplots 
for the top pathways suppressed and activated are highlighted. This plot depicts 
the enrichment scores (p values) and gene ratio as color and size, respectively.
Ridgeplots are also visualize enriched terms and their expression distributions 
of core enriched genes.

### Enrichment {.tabset}

#### CRISPR_52
```{r oragsea_run1, echo=FALSE, message=FALSE, warning=FALSE}
######################################################################
# run ORA/GSEA
######################################################################
treatment="CRISPR_52_with_IFNb"
main_gsea_ora_function_rna(cntrl_in=cntrl,
              treat_in=treatment,
              db_list,
              top_path_value=50,
              ORA_flag = "OFF",
              GSEA_flag = "ON")
```

#### CRISPR_53
```{r oragsea_run2, echo=FALSE, message=FALSE, warning=FALSE}
treatment="CRISPR_53_with_IFNb"
main_gsea_ora_function_rna(cntrl_in=cntrl,
              treat_in=treatment,
              db_list,
              top_path_value=50,
              ORA_flag = "OFF",
              GSEA_flag = "ON")
```