---
title: "CCBR1155_CS029758"
author: "Samantha Sevilla"
date: "`r Sys.Date()`"
output: html_document
editor_options: 
  chunk_output_type: console
---
# Background
######################################################################
## Project info
######################################################################
This analysis report includes:

| CS ID | Type | Species | CCBR | Notes | 
| -- | -- | -- | -- | -- |
| CS029758 | Cut and Run | Hg38 | 1155 |   |

| CS ID | Type | Notes | 
| -- | -- | -- |
| 53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb | broad | |
| 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb | broad | |


######################################################################
## Links to Tracks
######################################################################
USBC Genome Tracks were created for each project. They can be accessed here:

- [CS029758](https://hpcnihapps.cit.nih.gov/genome/cgi-bin/hgTracks?db=hg38&lastVirtModeType=default&lastVirtModeExtraState=&virtModeType=default&virtMode=0&nonVirtPosition=&position=chrX%3A15560138%2D15602945&hgsid=138544_oGancyaCdLajuCjbH3iYNL1RpiWC)

# Analysis
######################################################################
## Notes
######################################################################
- Deduplicated peaks were identified, using the following parameters:

  - fragments based method
  
  - normalized, relaxed method
  
- Peaks were then annotated with gene ID's. 

  - Comparisons including ALL genes include the corresponding gene of any signficant peaks identified (p value <0.05, fc +/- 1.5)
  
  - Peak annotations were stratified into those that were over represented, as well as those under-represented between each comparison. For example peakID chr1:100099186-100105441 had the following average counts for these two samples:	CRISPR_NC_2_Smyd3 (~65,287) and CRISPR_KO_10_Smyd3 (~349,940). The log2FoldChange between these two samples is approximately -2.29. The comparison CRISPR_NC_Smyd3_vs_KO_Smyd3 with a -2.29 log2FC would indicate that there is an UNDER-representation in CRISPR_NC_2_Smyd3 as compared to CRISPR_KO_10_Smyd3 for this gene ID (Cntnap5b).
  
  
```{r paths_lists, echo=FALSE, warning=FALSE, include=FALSE}
############################################################
# project information
############################################################
#paths
parent_dir="~/../../Volumes/CUTRUN/"
input_dir = "~/../../Volumes/CUTRUN/analysis/CS029758/"
output_car_dir = "~/../../Volumes/CUTRUN/analysis/CS029758/r_analysis_230321_newannotations/"
output_dir=output_car_dir

car_subpath=paste0(input_dir,"carlisle_230111/results/peaks/contrasts/")
bed_subpath=paste0(input_dir,"carlisle_230111/results/peaks/gopeaks/")

img_dir=paste0(output_car_dir,"img/")
dir.create(file.path(output_car_dir), showWarnings = FALSE)
dir.create(file.path(img_dir), showWarnings = FALSE)
cs_id="CS029758"
```

```{r user_params}
############################################################
# set global variables 
############################################################
padj_cutoff = 0.05
species = "Homo Sapiens"
log2fc_cutoff_car=0.3785 #FC of 1.3
method="fragments" #AUC or fragments
dedup_status="dedup"
scalesfbymean="Y"
rawcountsprescaled="N"
neg_control="HN6_igG_rabbit_1"
sample_consensus_threshold=0.1
read_minimum_threshold=10

gene_bodies_list=c("Promoter","Intron","Exon","5'UTR","3'UTR")
gene_bodies_filename=paste(unlist(substr(gene_bodies_list, start = 1, stop = 1)),collapse="")

# reference info
if (species=="Homo Sapiens"){
  txdb_pkg="TxDb.Hsapiens.UCSC.hg38.knownGene"
}

# group to ignore
ignore_groups=c()

```

```{r pkg, echo=FALSE, warning=FALSE, include=FALSE}
# package list
list.of.packages=c("DESeq2","edgeR", "tidyverse","dplyr", "DT","reshape2","pander","plotly", "ggplot2", "ggfortify", "ggrepel", "yaml", "EnhancedVolcano", "RUVSeq", "ChIPseeker", "stringr","RColorBrewer","HTSFilter","pheatmap",txdb_pkg, "UpSetR","ggupset","ggimage","grid","ggpubr","plyr","ggVennDiagram","biomaRt")

#install as needed
new.packages <- list.of.packages[!(list.of.packages %in% installed.packages()[,"Package"])]
if(length(new.packages)) BiocManager::install(new.packages)

# source functions
source(paste0(parent_dir,'github/CAR_core_functions.R'))

# load packages
invisible(lapply(list.of.packages, library, character.only = TRUE))

#negate
`%ni%` <- Negate(`%in%`)

# assign annodbs
if (species=="Homo Sapiens"){
  txdb<-TxDb.Hsapiens.UCSC.hg38.knownGene
  annodb="org.Hs.eg.db"
  genome="hg38"
} else if (species=="Mus Musculus"){
  genome="mm10"
}

```

## **Data Processing**
```{r contrast, include=FALSE,warning=FALSE, message=FALSE}
####################
# run metadata input
####################
#groups input
#awk '{print$1"_"$2,$1}' samples.tsv > groups.tab
#sed -i "s/ /\\t/g" groups.tab 
groups_df = read.csv(paste0(input_dir,"groups.tab"),sep="\t",header = FALSE)[,c("V1","V2")]
colnames(groups_df) = c("sampleid","group")
rownames(groups_df)=groups_df$sampleid

## if there is a subset gene list, select files
gene_file = paste0(parent_dir,"docs/gene_list_master.csv")
gene_df = read.csv(gene_file)
#read gene list
#gene_list=read.csv(paste0(parent_dir,"/docs/merged_gene_list_v2.csv"),sep=",")

# read in the contrast list
#cp carlisle_230111/config/contrasts.tsv .
contrast_df=read.csv(paste0(input_dir,"contrasts.tsv"),sep="\t")
colnames(contrast_df)=c("CNTRL","TREATMENT")
contrast_df$contrast=paste0(contrast_df$CNTRL,"_vs_",contrast_df$TREATMENT)

#filter
groups_df=subset(groups_df, group %ni% ignore_groups)
contrast_df=subset(contrast_df, (CNTRL %in% unique(groups_df$group)) | (TREATMENT %in% unique(groups_df$group)))
```

The following samples are included in this analysis:
```{r dt_samp, echo=FALSE,warning=FALSE, message=FALSE}
DT::datatable(groups_df)
```

The following contrasts are included in this analysis:
```{r dt_group, echo=FALSE,warning=FALSE, message=FALSE}
DT::datatable(contrast_df)
```

## **Quality Control**
Raw counts are read in and filtered by counts per million reads (CPM). Filtering thresholds are set such that at least two samples must have a CPM of >0.5 to be reviewed. Two plots were generated for each sample: Relative log expression (RLE) plots and Principal Coordinate Analysis (PCA) plots.

Reviewing sample QC allows us to determine if there are outliers present in analysis and the success of data normalization. If there are, samples are removed in the "Selected Samples Normalized" plots.

## Quality Control  {.tabset}

#### 53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb
```{r qc1, echo=FALSE, warning=FALSE, message=FALSE}
peak_anno=main_prep_qc_core(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                            peak_type="broadGo_peaks")
chipseeker_plots(peak_anno)
```

#### 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb
```{r qc2, echo=FALSE, warning=FALSE, message=FALSE}
peak_anno=main_prep_qc_core(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                            peak_type="broadGo_peaks")
chipseeker_plots(peak_anno)
```

## **Data Processing**
Prepare peak tables for both significantly differentiated and non-significantly differentiated peaks
```{r collapsedf, echo=FALSE, warning=FALSE, message=FALSE}
# create significant peak tables per contrast
create_sig_contrast_df(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                       peak_type="broadGo_peaks",
                       gene_list_name=c("APM","IFNalpha"))

rna_path=paste0(parent_dir,"analysis/CS028891/r_analysis_230323_newannotation/",
                "contrast_sig_CRISPR_53_with_IFNb-parental_HN6_with_IFNb.csv")
create_sig_contrast_df(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                       peak_type="broadGo_peaks",
                       gene_list_name=c("APM","IFNalpha"),
                       rna_seq_path=rna_path)

# collapse peaks per sample
create_collapsed_sample_df(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb")
create_collapsed_sample_df(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb")

# collapse significant peak tables per contrast to annotation level
create_collapsed_contrast_df(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb")
create_collapsed_contrast_df(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb")
```

## **Peak Distributions**

Create pie charts per sample, per contrast, and differentiated by +/- log2FC

### 53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb {.tabset}
```{r, pie1, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_piecharts(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb")
```

### 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb {.tabset}
```{r, pie2, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_piecharts(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb")
```

## **Sample Heatmap**

### 53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb {.tabset}
```{r, sh1, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
sample_sub_list=c("HN6_H3K27me3_IFNb_1","HN6_H3K27me3_IFNb_2",
                  "53_H3K27me3_IFNb_1","53_H3K27me3_IFNb_2")
# ALL GENES
generate_replicate_heatmaps(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                            scale_flag="ON",
                            sample_subset=sample_sub_list,
                            gene_list_name="")

# subsets of gene lists
gene_list_options=c("APM_INFA")
for (genelistID in gene_list_options){
  gene_list_name=genelistID
  generate_replicate_heatmaps(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                            scale_flag="ON",
                            gene_list_name=genelistID,
                            sample_subset=sample_sub_list)
}
```

### 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb {.tabset}
```{r, sh2, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
sample_sub_list=c("HN6_H4K20me3_IFNb_2","HN6_H4K20me3_IFNb_3",
                  "53_H4K20me3_IFNb_2","53_H4K20me3_IFNb_3")

# ALL GENES
generate_replicate_heatmaps(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                           scale_flag="ON",
                           sample_subset=sample_sub_list,
                           gene_list_name="")

# subsets of gene lists
gene_list_options=c("APM_INFA")
for (genelistID in gene_list_options){
  generate_replicate_heatmaps(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                            scale_flag="ON",
                            gene_list_name=genelistID,
                            sample_subset=sample_sub_list)
}
# ALL GENES with MRNA data
generate_replicate_heatmaps(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                           scale_flag="ON",
                           sample_subset=sample_sub_list,
                           rna_flag="Y",
                           gene_list_name="")

# subsets of gene lists
gene_list_options=c("APM_INFA")
for (genelistID in gene_list_options){
  generate_replicate_heatmaps(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                            scale_flag="ON",
                            gene_list_name=genelistID,
                           rna_flag="Y",
                            sample_subset=sample_sub_list)
}
```

## **Volcano Plots**

Volcano plots are created to show the statistical significance (FDR) versus magnitude of change (fold change).

### 53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb {.tabset}
```{r, v1, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_volcano_plots(data_type="CAR",
                       contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                       gene_list_name=c("APM","IFNalpha"))
```

### 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb {.tabset}
```{r, v2, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_volcano_plots(data_type="CAR",
                       contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                       gene_list_name=c("APM","IFNalpha"))
```

## *Intensity Boxplots*

### 53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb {.tabset}
```{r, box1, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_intensity_boxplot(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                           sig_flag="N",
                           gene_list_name="",
                           scale_factor="Y")

generate_intensity_boxplot(contrast_id="53_H3K27me3_IFNb_vs_HN6_H3K27me3_IFNb",
                           sig_flag="Y",
                           gene_list_name="APM_INFA",
                           scale_factor="Y")
```

### 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb {.tabset}
```{r, box2, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
generate_intensity_boxplot(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                           sig_flag="N",
                           gene_list_name="",
                           scale_factor="Y")

generate_intensity_boxplot(contrast_id="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                           sig_flag="N",
                           gene_list_name="APM_INFA",
                           scale_factor="Y")
```

### Combo of 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb with CS028891 53_INFB_SMYD3_vs_HN6_INFB_SMYD3
### run with and without distal
```{r, box3, echo=FALSE, warning=FALSE,message=FALSE,fig.height=7}
csid2_path=paste0(parent_dir,"analysis/CS031308/r_analysis_230307/")

gene_bodies_list=c("Promoter","Intron","Exon","5'UTR","3'UTR","Distal")
gene_bodies_filename=paste(unlist(substr(gene_bodies_list, start = 1, stop = 1)),collapse="")
generate_intensity_boxplot_two_projs(contrast_id1="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                                     csid2="CS031308",
                                     csid2_path=csid2_path,
                                     contrast_id2="53_INFB_SMYD3_vs_HN6_INFB_SMYD3",
                                     gene_list_name="")
generate_intensity_boxplot_two_projs(contrast_id1="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                                     csid2="CS031308",
                                     csid2_path=csid2_path,
                                     contrast_id2="53_INFB_SMYD3_vs_HN6_INFB_SMYD3",
                                     gene_list_name="APM_INFA")
                                     
gene_bodies_list=c("Promoter","Intron","Exon","5'UTR","3'UTR")
gene_bodies_filename=paste(unlist(substr(gene_bodies_list, start = 1, stop = 1)),collapse="")
generate_intensity_boxplot_two_projs(contrast_id1="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                                     csid2="CS031308",
                                     csid2_path=csid2_path,
                                     contrast_id2="53_INFB_SMYD3_vs_HN6_INFB_SMYD3",
                                     gene_list_name="")
generate_intensity_boxplot_two_projs(contrast_id1="53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                                     csid2="CS031308",
                                     csid2_path=csid2_path,
                                     contrast_id2="53_INFB_SMYD3_vs_HN6_INFB_SMYD3",
                                     gene_list_name="APM_INFA")
```


## **VENN DIAGRAMS **

### Combo between 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb and CS031308 53_INFB_SMYD3_vs_HN6_INFB_SMYD3
```{r}
create_venn_diagrams_mult_projs(cs_id_list=c("CS029758","CS031308"),
                                cs_id_path_list=c(output_dir,
                                                  paste0(parent_dir,"analysis/CS031308/r_analysis_230307/")),
                                contrast_list=c("53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                                                "53_INFB_SMYD3_vs_HN6_INFB_SMYD3"),
                                sig_filter_list=c("Y","N"),
                                gene_list_name="APM_INFA",
                                gene_list_to_include="APM_IFNalpha"
)
```

### Combo between 53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb and CS031308 53_INFB_SMYD3_vs_HN6_INFB_SMYD3 
### AND CS0329689 5-3_H3K9me3_IFNb_vs_HN6_H3K9me3_IFNb

```{r}
create_venn_diagrams_mult_projs(cs_id_list=c("CS029758","CS031308","CS029689"),
                                cs_id_path_list=c(output_dir,
                                                  paste0(parent_dir,
                                                         "analysis/CS031308/r_analysis_230307/"),
                                                  paste0(parent_dir,
                                                         "analysis/CS029689/r_analysis_230326/")),
                                contrast_list=c("53_H4K20me3_IFNb_vs_HN6_H4K20me3_IFNb",
                                                "53_INFB_SMYD3_vs_HN6_INFB_SMYD3",
                                                "53_H3K4me3_IFNb_vs_HN6_H3K4me3_IFNb"),
                                sig_filter_list=c("N","N","N"),
                                gene_list_name="APM_INFA",
                                gene_list_to_include="APM_IFNalpha"
)
```
